nfs/src/rpc_task.cpp

```diff
@@ -476,6 +476,377 @@ void rpc_task::run_setattr()
+                 */
+                for (size_t j = 0; j <= i; j++) {
+                    readfile_handle->release(
+                        bytes_vector[j].offset,
+                        bytes_vector[j].length);
+                }
+
+                goto read_from_cache;
+            }
+        }
+        else
+        {
+            // This indicates that some data is missing in the cache.
+            break;
+        }
+    }
+
+    /*
+     * Hold an extra ref since we do not want to send the response
+     * before all the reads complete.
+     * It is ok to access this without a lock since this is the only thread
+     * at this point which will access this.
+     */
+    num_of_reads_issued_to_backend = 1;
+
+    for (size_t i = 0; i < size; i++) {
+        /*
+         * One or more chunks don't have the requested data, issue read
+         * for those.
+         */
+        if (!bytes_vector[i].get_membuf()->is_uptodate())
+        {
+            readfile_from_server(bytes_vector[i]);
+        }
+    }
+
+    // Drop the extra ref that we held under the lock.
+    {
+        std::unique_lock<std::shared_mutex> lock(read_task_lock);
+        --num_of_reads_issued_to_backend;
+
+        if (num_of_reads_issued_to_backend == 0) {
+            goto send_response;
+        }
+        else
+        {
+            return;
+        }
+    }// End of lock
+
+read_from_cache:
+    AZLogDebug("Data read from cache, also releasing buffer."
+               " offset: {}, size {}",
+               rpc_api.read_task.get_offset(),
+               rpc_api.read_task.get_size());
+
+send_response:
+    assert (num_of_reads_issued_to_backend == 0);
+
+    // Send the response.
+    send_readfile_response(0 /* success status */);
+}
+
+void rpc_task::send_readfile_response(int status)
+{
+    /*
+     * We should have completed all the reads before sending the response
+     * to caller.
+     */
+    assert(num_of_reads_issued_to_backend == 0);
+
+    if (status != 0) {
+        // Non-zero status indicates failure, reply with error in such cases.
+        reply_error(status);
+        return;
+    }
+
+    const size_t count = bytes_vector.size();
+
+    // Create an array of iovec struct
+    struct iovec iov[count];
+
+    // Fetch the caller requested size.
+    const size_t req_size = rpc_api.read_task.get_size();
+    size_t remaining_size = req_size;
+
+    for (size_t i = 0; (i < count && remaining_size > 0); i++)
+    {
+        if (remaining_size >= bytes_vector[i].length) {
+            /*
+             * If the first chunk itself is empty, then there is no need to
+             * look further, so just send empty response as we reach here only
+             * in the case of success read but bytes read is 0.
+             */
+            if ((i==0) && (bytes_vector[i].length == 0))
+            {
+                reply_iov(nullptr, 0);
+                return;
+            }
```

Comment:
what are you trying to handle?
note that fuse cannot handle partial read responses. So if libnfs can return partial read response we need to issue subsequent read calls till we get all the data we need or we get eof.